packages/@react-spectrum/s2/src/Table.tsx

```diff
@@ -11,85 +11,1079 @@
+  }), [isQuiet, density, overflowMode, loadingState, isInResizeMode, setIsInResizeMode]);
+
+  let isLoading = loadingState === 'loading' || loadingState === 'loadingMore';
+  let scrollRef = useRef(null);
+  let memoedLoadMoreProps = useMemo(() => ({
+    isLoading: isLoading,
+    onLoadMore
+  }), [isLoading, onLoadMore]);
+  useLoadMore(memoedLoadMoreProps, scrollRef);
+  let isCheckboxSelection = props.selectionMode === 'multiple' || props.selectionMode === 'single';
+
+  return (
+    <ResizableTableContainer
+      onResize={propsOnResize}
+      onResizeEnd={onResizeEnd}
+      onResizeStart={onResizeStart}
+      className={(UNSAFE_className || '') + mergeStyles(tableWrapper, styles)}
+      style={UNSAFE_style}>
+      <UNSTABLE_Virtualizer layout={layout}>
+        <InternalTableContext.Provider value={context}>
+          <RACTable
+            ref={scrollRef}
+            style={{WebkitTransform: 'translateZ(0)'}}
+            className={renderProps => table({
+              ...renderProps,
+              scale,
+              isCheckboxSelection,
+              isQuiet
+            })}
+            selectionBehavior="toggle"
+            onRowAction={onAction}
+            {...otherProps} />
+        </InternalTableContext.Provider>
+      </UNSTABLE_Virtualizer>
+    </ResizableTableContainer>
+  );
+}
+
+const centeredWrapper = style({
+  display: 'flex',
+  alignItems: 'center',
+  justifyContent: 'center',
+  width: 'full',
+  height: 'full'
+});
+
+export interface TableBodyProps<T> extends Omit<RACTableBodyProps<T>, 'style' | 'className' | 'dependencies'> {}
+
+export function TableBody<T extends object>(props: TableBodyProps<T>) {
+  let {items, renderEmptyState, children} = props;
+  let {loadingState} = useContext(InternalTableContext);
+  let emptyRender;
+  let renderer = children;
+  let stringFormatter = useLocalizedStringFormatter(intlMessages, '@react-spectrum/s2');
+  let loadMoreSpinner = (
+    <UNSTABLE_TableLoadingIndicator className={style({height: 'full', width: 'full'})}>
+      <div className={centeredWrapper}>
+        <ProgressCircle
+          isIndeterminate
+          aria-label={stringFormatter.format('table.loadingMore')} />
+      </div>
+    </UNSTABLE_TableLoadingIndicator>
+  );
+
+  // If the user is rendering their rows in dynamic fashion, wrap their render function in Collection so we can inject
+  // the loader. Otherwise it is a static renderer and thus we can simply add the table loader after
+  // TODO: this assumes that the user isn't providing their children in some wrapper though and/or isn't doing a map of children
+  // (though I guess they wouldn't provide items then so the check for this is still valid in the latter case)...
+  if (typeof children === 'function' && items) {
+    renderer = (
+      <>
+        <Collection items={items}>
+          {children}
+        </Collection>
+        {loadingState === 'loadingMore' && [...items].length > 0 && loadMoreSpinner}
```

Comment:
the length check here is to prevent a situation where somehow the `loadingState = "loadingMore"' but no items are actually loaded aka the below:
![image](https://github.com/user-attachments/assets/bf5d92b0-1308-43cd-86a6-480e2ab02205)

For `loadingState = "loading"` we don't need these checks because the rendering of that circle/empty state is provided to `renderEmptyState` instead so RAC table handles rendering it only when the collection size is 0.

We could consider the "loadingMore" case a user error and get rid of the length check perhaps, what do you think? Otherwise I need some way of knowing the collection is empty...

Reply:
Hmm yeah I'd consider that programmer error I think